### DISCUSSION TITLE
High process memory in quorum queue

### DISCUSSION THREAD

**Discussion by annieblomgren** [rabbitmq/rabbitmq-server#3063] [Questions]
Hi,

RabbitMQ version 3.8.16
Erlang 23.2.3
Environment="RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS=+sbwt none +sbwtdcpu none +sbwtdio none"

Perftest version 2.11.0 with command `​bin/runjava com.rabbitmq.perf.PerfTest -u "testqq2" --size 10 --producers 5 --consumers 0 -f persistent -H $URL --auto-delete=false --queue-args x-queue-type=quorum -pmessages 400000`

A quorum queue with 2M messages, no publishers nor consumers is having around 1.6GB in process memory.
After a full GC process memory goes down to around 1.4GB

<img width="449" alt="Screen Shot 2021-05-24 at 17 22 02" src="https://user-images.githubusercontent.com/48422873/119369633-83968580-bd08-11eb-96dc-5135ef81551a.png">

<img width="907" alt="Screen Shot 2021-05-24 at 16 35 33" src="https://user-images.githubusercontent.com/48422873/119368703-74fb9e80-bd07-11eb-8738-de15a3147452.png">
<img width="803" alt="Screen Shot 2021-05-24 at 16 54 12" src="https://user-images.githubusercontent.com/48422873/119368707-75943500-bd07-11eb-95f0-2e9a415e609f.png">

The same number of messages for a classic queue sits around 300 MiB in process memory before a full GC, and ~60MiB after.  

Have changed `raft.wal_max_size_bytes` and `x-max-in-memory-bytes` to low values but no difference. 

Grafana dashboards before and after GC on QQ:
​https://snapshot.raintank.io/dashboard/snapshot/IphlxnoFeXsjqw6wVH0G0vYS2svJ2GFJ
​https://snapshot.raintank.io/dashboard/snapshot/PhdZV2Y2Z3oMyxbuft8f8nLs72vvcWL1

**Question: what is taking up so much process memory for QQs, or is this to be expected?**

**Comment by kjnilsson** [comment]
Quorum queues keep all messages in memory as well as on disk by default.
Try configuring a max-in-memory length limit to make quorum queues re-read
messages from it's raft log when they are consumed instead. Not this has a
consumption performance penalty.

https://www.rabbitmq.com/quorum-queues.html#memory-limit

Cheers
Karl

On Mon, 24 May 2021 at 16:26, annieblomgren ***@***.***>
wrote:

> Hi,
>
> RabbitMQ version 3.8.16
> Erlang 23.2.3
> Environment="RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS=+sbwt none +sbwtdcpu none
> +sbwtdio none"
>
> Perftest version 2.11.0 with command ​bin/runjava
> com.rabbitmq.perf.PerfTest -u "testqq2" --size 10 --producers 5 --consumers
> 0 -f persistent -H $URL --auto-delete=false --queue-args
> x-queue-type=quorum -pmessages 400000
>
> A quorum queue with 2M messages, no publishers nor consumers is having
> around 1.6GB in process memory.
> After a full GC process memory goes down to around 1.4GB
>
> [image: Screen Shot 2021-05-24 at 17 22 02]
> <https://user-images.githubusercontent.com/48422873/119369633-83968580-bd08-11eb-96dc-5135ef81551a.png>
>
> [image: Screen Shot 2021-05-24 at 16 35 33]
> <https://user-images.githubusercontent.com/48422873/119368703-74fb9e80-bd07-11eb-8738-de15a3147452.png>
>
> [image: Screen Shot 2021-05-24 at 16 54 12]
> <https://user-images.githubusercontent.com/48422873/119368707-75943500-bd07-11eb-95f0-2e9a415e609f.png>
>
> The same number of messages for a classic queue sits around 300 MiB in
> process memory before a full GC, and ~60MiB after.
>
> Have changed raft.wal_max_size_bytes and x-max-in-memory-bytes to low
> values but no difference.
>
> Grafana dashboards before and after GC on QQ:
> ​
> https://snapshot.raintank.io/dashboard/snapshot/IphlxnoFeXsjqw6wVH0G0vYS2svJ2GFJ
> ​
> https://snapshot.raintank.io/dashboard/snapshot/PhdZV2Y2Z3oMyxbuft8f8nLs72vvcWL1
>
> *Question: what is taking up so much process memory for QQs, or is this to
> be expected?*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/3063>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFDECNBE2I5P72R2EQLTPJVYRANCNFSM45NN54EQ>
> .
>


-- 
*Karl Nilsson*

**Reply by johanrhodin** [reply]
OK I tried with `max-in-memory-length` and it did free up some process memory but not much (from ~1.6GB to ~1.4GB). 
![rmq_process_memory_max-in-memory-length](https://user-images.githubusercontent.com/460369/119376721-e729a000-bc81-11eb-9510-1272f10d97c0.png)

**Comment by kjnilsson** [comment]
If you applied the policy to a queue with messages already in memory as you
can see it will still keep them in memory until they are consumed. Try the
test starting with an empty queue and the policy in place.

On Mon, 24 May 2021 at 17:21, Johan Rhodin ***@***.***> wrote:

> OK I tried with max-in-memory-length and it did free up some process
> memory but not much (from ~1.6GB to ~1.4GB).
> [image: rmq_process_memory_max-in-memory-length]
> <https://user-images.githubusercontent.com/460369/119376721-e729a000-bc81-11eb-9510-1272f10d97c0.png>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/3063#discussioncomment-777489>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFDTEEYKK4MK434L2VTTPJ4HLANCNFSM45NN54EQ>
> .
>


-- 
*Karl Nilsson*

**Comment by johanrhodin** [comment]
OK, that brought down process memory to 478MiB 👍.
<img width="1018" alt="max-in-memory-from-the-start" src="https://user-images.githubusercontent.com/460369/119380448-5903e880-bc86-11eb-840e-11116c311705.png">
Is it possible to reduce it even further? As a comparison a (deprecated) mirrored classic queue with the same 2M messages and 1 replica is using ~2.5MiB process memory.

And for cluster where the user has already loaded in the message before applying the policy, is there a way to tell RabbitMQ to to page out those message (similar to changing `vm_memory_high_watermark_paging_ratio` for classic queues).

**Reply by kjnilsson** [reply]
Quorum queues need to keep certain accounting data in memory for efficiency and correctness so now right now it is not possible to reduce memory use further. We could consider reducing memory when the policy comes into place though.

**Reply by kjnilsson** [reply]
We do have some brilliant changes coming for lazy non replicated queues however which will use even less memory. Also if you need replication _and_ low memory the upcoming streams provide this but of course they aren't really queues.

**Reply by lhoguin** [reply]
Both lazy and non-lazy. :-) Maybe I will look into QQs after that, 478MiB sounds just a little bit much.

If you know how I'd be interested to know what this returns, where `QPid` is the pid of the queue process (Process column in RabbitMQ top): `process_info(QPid, garbage_collection_info)`.

**Reply by kjnilsson** [reply]
I've done a bit of research today and come up with a couple of ways to reduce QQ memory use by ~50-60% for a queue with 2M messages and x-max-in-memory-length = 0. That does still leave ~200MiB in this case.

**Reply by johanrhodin** [reply]
This is what the `garbage_collection_info` looks like:
```
{garbage_collection_info,[{old_heap_block_size,55185655},
                          {heap_block_size,7427328},
                          {mbuf_size,50640},
                          {recent_size,36970},
                          {stack_size,11},
                          {old_heap_size,44517067},
                          {heap_size,2660543},
                          {bin_vheap_size,0},
                          {bin_vheap_block_size,46422},
                          {bin_old_vheap_size,36},
                          {bin_old_vheap_block_size,46422}]}
```
Attaching the full `recon:info` and `recon:get_state`, too.
[recon_info.txt](https://github.com/rabbitmq/rabbitmq-server/files/6541076/recon_info.txt)
[recon_get_state.gz](https://github.com/rabbitmq/rabbitmq-server/files/6541086/recon_get_state.gz)

**Reply by lhoguin** [reply]
Most of that memory can be GC @kjnilsson there's a lot of room for improvement. Everything in "old" would get to 0 on a full GC and that's 80%+ of the memory usage. Either QQ is creating a lot of garbage or it is not doing GC often enough (or both). I'm happy to look into it after I am done with the new CQ index, if you want.

**Reply by kjnilsson** [reply]
that doesn't sound quite right as I did a test with 2M messages + a full GC and it remained at ~470MiB

**Reply by lhoguin** [reply]
Right I skipped a step in my analysis, "old" would get to 0 but "new" would increase after the full GC. In their case I would expect about 120MB to be freed (and not 400+ as I initially thought), but those would probably be used again almost immediately, or kept around as unused memory for future small GCs. I don't know about your case though. I would try with `fullsweep_after` set to 0 to see what happens.

Clearly though the biggest win to be had is to reduce the size of the state so I would go with your initial idea above first.
